src/builder/QuarkOperationHelper.sol

```diff
@@ -0,0 +1,150 @@
+                    currentIndex[j]++;
+                    break;
+                }
+            }
+        }
+
+        // Create new arrays for merged operations and actions
+        IQuarkWallet.QuarkOperation[] memory mergedQuarkOperations = new IQuarkWallet.QuarkOperation[](uniqueChainCount);
+        Actions.Action[] memory mergedActions = new Actions.Action[](uniqueChainCount);
+
+        // Merge operations for each unique chain
+        for (uint256 i = 0; i < uniqueChainCount; ++i) {
+            if (operationCounts[i] == 1) {
+                // If there's only one operation for this chain, we don't need to merge
+                mergedQuarkOperations[i] = groupedQuarkOperations[i][0];
+                mergedActions[i] = groupedActions[i][0];
+            } else {
+                // Merge multiple operations for this chain
+                (mergedQuarkOperations[i], mergedActions[i]) =
+                    mergeOperations(groupedQuarkOperations[i], groupedActions[i]);
+            }
+        }
+
+        return (mergedQuarkOperations, mergedActions);
+    }
+
+    // Note: Assumes all the quark operations are for the same quark wallet.
+    function mergeOperations(IQuarkWallet.QuarkOperation[] memory quarkOperations, Actions.Action[] memory actions)
+        internal
+        pure
+        returns (IQuarkWallet.QuarkOperation memory, Actions.Action memory)
+    {
+        address[] memory callContracts = new address[](quarkOperations.length);
+        bytes[] memory callDatas = new bytes[](quarkOperations.length);
+        bytes[] memory scriptSources = new bytes[](quarkOperations.length + 1);
+
+        for (uint256 i = 0; i < quarkOperations.length; ++i) {
+            callContracts[i] = quarkOperations[i].scriptAddress;
+            callDatas[i] = quarkOperations[i].scriptCalldata;
+            // Note: For simplicity, we assume there is one script source per quark operation for now
+            scriptSources[i] = quarkOperations[i].scriptSources[0];
+        }
+        scriptSources[quarkOperations.length] = type(Multicall).creationCode;
+
+        bytes memory multicallCalldata = abi.encodeWithSelector(Multicall.run.selector, callContracts, callDatas);
+
+        // Construct QuarkOperation and Action
+        // Note: We give precedence to the last operation and action for now because any earlier operations
+        // are auxiliary (e.g. wrapping an asset)
+        IQuarkWallet.QuarkOperation memory lastQuarkOperation = quarkOperations[quarkOperations.length - 1];
+        IQuarkWallet.QuarkOperation memory mergedQuarkOperation = IQuarkWallet.QuarkOperation({
+            nonce: lastQuarkOperation.nonce,
+            scriptAddress: CodeJarHelper.getCodeAddress(type(Multicall).creationCode),
+            scriptCalldata: multicallCalldata,
+            scriptSources: scriptSources,
+            expiry: lastQuarkOperation.expiry
+        });
+
+        return (mergedQuarkOperation, actions[actions.length - 1]);
```

Comment:
do we discard all but the last action?

Reply:
Correct. As of now, that should only be the wrap actions, which are fine to discard. In the long run, we'll likely want to support multiple action contexts for a single quark operation. Right now, they are 1:1